File: api-server/app/schemas/paper_trading.py
"""
Paper Trading Pydantic 스키마
"""
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional, Dict, Any, List


class PaperTradingSessionCreate(BaseModel):
    name: Optional[str] = None
    symbol: str = Field(..., example="BTC/USDT")
    strategy: str = Field(..., example="Statistical Arbitrage")
    initial_capital: float = Field(..., gt=0, example=10000.0)
    settings: Optional[Dict[str, Any]] = None


class PaperTradingSessionUpdate(BaseModel):
    status: Optional[str] = None
    current_balance: Optional[float] = None
    current_equity: Optional[float] = None
    total_trades: Optional[int] = None
    winning_trades: Optional[int] = None
    losing_trades: Optional[int] = None
    total_pnl: Optional[float] = None
    total_commission: Optional[float] = None
    max_equity: Optional[float] = None
    max_drawdown: Optional[float] = None


class PaperTradingSessionResponse(BaseModel):
    id: int
    name: Optional[str]
    symbol: str
    strategy: str
    initial_capital: float
    current_balance: float
    current_equity: float
    status: str
    start_time: datetime
    end_time: Optional[datetime]
    last_update: datetime
    settings: Optional[Dict[str, Any]]
    total_trades: int
    winning_trades: int
    losing_trades: int
    total_pnl: float
    total_commission: float
    max_equity: Optional[float]
    max_drawdown: float

    class Config:
        orm_mode = True


class PaperTradingTradeResponse(BaseModel):
    id: int
    session_id: int
    symbol: str
    side: str
    position_side: str
    quantity: float
    entry_price: Optional[float]
    exit_price: Optional[float]
    commission: float
    slippage: float
    pnl: Optional[float]
    pnl_percent: Optional[float]
    position_type: Optional[str]
    stop_loss: Optional[float]
    take_profit: Optional[float]
    exit_reason: Optional[str]
    timestamp: datetime

    class Config:
        orm_mode = True


class PaperTradingSnapshotResponse(BaseModel):
    id: int
    session_id: int
    balance: float
    equity: float
    unrealized_pnl: float
    open_positions: Optional[List[Dict[str, Any]]]
    market_price: Optional[float]
    timestamp: datetime

    class Config:
        orm_mode = True


class PaperTradingStats(BaseModel):
    session_id: int
    total_trades: int
    winning_trades: int
    losing_trades: int
    win_rate: float
    total_pnl: float
    total_commission: float
    total_return: float
    max_drawdown: float
    sharpe_ratio: Optional[float]
    sortino_ratio: Optional[float]


class KlineData(BaseModel):
    time: int = Field(..., description="Unix timestamp in milliseconds")
    open: float
    high: float
    low: float
    close: float
    volume: float

File: web-dashboard/src/types/paper_trading.ts
/**
 * Paper Trading TypeScript Types
 * Based on backend schemas: app/schemas/paper_trading.py
 */

export type SessionStatus = "active" | "paused" | "stopped" | "completed";

export interface PaperTradingSession {
  id: number;
  name: string | null;
  symbol: string;
  strategy: string;
  initial_capital: number;
  current_balance: number;
  current_equity: number;
  status: SessionStatus;
  start_time: string; // ISO 8601 datetime
  end_time: string | null; // ISO 8601 datetime
  last_update: string; // ISO 8601 datetime
  settings: Record<string, any> | null;
  total_trades: number;
  winning_trades: number;
  losing_trades: number;
  total_pnl: number;
  total_commission: number;
  max_equity: number | null;
  max_drawdown: number;
}

export interface PaperTradingSessionCreate {
  name?: string | null;
  symbol: string;
  strategy: string;
  initial_capital: number;
  settings?: Record<string, any> | null;
}

export interface PaperTradingTrade {
  id: number;
  session_id: number;
  symbol: string;
  side: string; // "buy" | "sell"
  position_side: string; // "long" | "short"
  quantity: number;
  entry_price: number | null;
  exit_price: number | null;
  commission: number;
  slippage: number;
  pnl: number | null;
  pnl_percent: number | null;
  position_type: string | null;
  stop_loss: number | null;
  take_profit: number | null;
  exit_reason: string | null;
  timestamp: string; // ISO 8601 datetime
}

export interface PaperTradingSnapshot {
  id: number;
  session_id: number;
  balance: number;
  equity: number;
  unrealized_pnl: number;
  open_positions: Array<Record<string, any>> | null;
  market_price: number | null;
  timestamp: string; // ISO 8601 datetime
}

export interface PaperTradingStats {
  session_id: number;
  total_trades: number;
  winning_trades: number;
  losing_trades: number;
  win_rate: number; // percentage
  total_pnl: number;
  total_commission: number;
  total_return: number; // percentage
  max_drawdown: number;
  sharpe_ratio: number | null;
  sortino_ratio: number | null;
}

export interface KlineData {
  time: number; // Unix timestamp in milliseconds
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}
